### ISSUE TITLE
[qt5.11] Pen.stringCenteredIn + horizontal scrolling is super creepy

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#3822]
I'm quite sure, with the older Qt bindings, that stringCenteredIn exhibits the expected behavior: always use the given bounds and, at the edge of a ScrollView, simply clip the display so that the text appears to be normal, but you just can't see the portion that is outside the window's physical bounds.

With Qt 5.11, stringCenteredIn now clips the *bounds* (Rect) to the visible area and then -- never let it be said the Qt developers have no sense of humor -- squeezes the text to fit within the bounds, at the cost of legibility.

I suppose possibly this is just a(nother) Qt bug we might have to live with, but perhaps there is some property we can set in the backend to prevent this behavior.

```
(
w = Window("hscroll", Rect(800, 200, 500, 400));
v = ScrollView(w, w.view.bounds.insetBy(4, 4)).background_(Color(0.8, 0.8, 1));
Array.fill(8, { |i|
	UserView(v, Rect(4 + (i * 110), 4, 106, 20))
	.background_(Color(
		i.linlin(0, 7, 0.2, 0.6),
		i.linlin(0, 7, 0.5, 0.1),
		i.linlin(0, 7, 0.1, 0.3)
	))
	.drawFunc_({ |view|
		Pen.color_(Color.white)
		.stringCenteredIn("Text #" ++ i, view.bounds.moveTo(0, 0))
	});
});
w.front;
)
```

**Comment by dyfer** [comment]
I can confirm the same happens when building agains QT 5.9.1 (on macOS), while it does not happen on an older build agains QT 5.5.
FWIW this scaling takes advantage of a highDPI display, so it looks kind of legible (if very funny!) up until ~50% of the original width.

**Comment by scztt** [comment]
This isn't related to text, it's related to how we paint UserView onto the actual widget:
`if( _paint ) painter.drawPixmap( e->rect(), _pixmap );` (see QcCanvas.cpp)
In this case, e->rect() is the actual area being painted, which may not match the actual pixmap / widget size (in the bug case, because it's offscreen).
There's another version of `drawPixmap` that has a source rect as an additional argument - we probably need to use that, and pass in a source rect that's ONLY the part overlapping with e->rect(). I think the fix looks something like this (with pixelRatio scaling):
```	auto sourceRect = e->rect().intersected(QRect(0, 0, size().width(), size().height()));
	sourceRect = QRect(sourceRect.topLeft() * pixelRatio, sourceRect.size() * pixelRatio);
	if( _paint ) painter.drawPixmap( e->rect(), _pixmap, sourceRect);
```
Can someone test that on a system that isn't hi-dpi?

**Comment by snappizz** [comment]
@scztt this worked for me:

```
  if ( _paint ) {
    int pixelRatio = round(backingStore()->window()->devicePixelRatio());
    auto sourceRect = e->rect().intersected(QRect(0, 0, size().width(), size().height()));
    sourceRect = QRect(sourceRect.topLeft() * pixelRatio, sourceRect.size() * pixelRatio);
    painter.drawPixmap( e->rect(), _pixmap, sourceRect);
  }
```

**Comment by scztt** [comment]
Is your pix ratio 1? That looks a little wrong to me, only because I think that painter might be in logical coords and the pixnap is definitely in pixels.

**Comment by snappizz** [comment]
yes, it is 1.

**Comment by snappizz** [comment]
@scztt by "looks a little wrong," do you mean the code seems incorrect, or it looks wrong on hi-dpi?

**Comment by snappizz** [comment]
fixed in #3867
